layers/vk_api_hash.h

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2023 Nintendo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <functional>
+
+#include <vulkan/vulkan.h>
+
+namespace std {
+
+#define APPEND_HASH(var) res = res * 31 + std::hash<std::decay<decltype(var)>::type>()(var)
+
+#define APPEND_HASH_ARRAY(array_member, current_length_member, max_length) \
+    APPEND_HASH(current_length_member);                                    \
+    for (size_t i = 0; i < current_length_member; ++i) {                   \
+        APPEND_HASH((array_member)[i]);                                    \
+    }                                                                      \
+    for (size_t i = current_length_member; i < max_length; ++i) {          \
+        APPEND_HASH(0u);                                                   \
+    }
+
+template <>
+struct hash<VkStencilOpState> {
+    std::size_t operator()(VkStencilOpState const& o) const {
+        size_t res = 17;
```

Comment:
Similarly an explanation of the 17 constant that appears many times below here. Probably it should be a constexpr and be defined higher so you can have a single explanation of its purpose.